module/userExtension/helper.js

```diff
@@ -211,304 +213,391 @@ module.exports = class UserExtensionHelper {
-                ].programs.map((program) => program.toString());
-
-                if (userRole.programOperation == 'OVERRIDE') {
-                  existingUser.platformRoles[userPlatformRoleToUpdate].programs = programIds;
-                } else if (userRole.programOperation == 'APPEND' || userRole.programOperation == 'ADD') {
-                  existingUser.platformRoles[userPlatformRoleToUpdate].programs =
-                    existingUser.platformRoles[userPlatformRoleToUpdate].programs.concat(programIds);
-                  existingUser.platformRoles[userPlatformRoleToUpdate].programs = _.uniq(
-                    existingUser.platformRoles[userPlatformRoleToUpdate].programs,
-                  );
-                } else if (userRole.programOperation == 'REMOVE') {
-                  if (programIds.length > 0) {
-                    programIds.forEach((programId) => {
-                      _.pull(existingUser.platformRoles[userPlatformRoleToUpdate].programs, programId.toString());
-                    });
+                  if (!programId) {
+                    userRole.status = messageConstants.apiResponses.PROGRAM_NOT_FOUND;
+                    userRolesUploadedData.push(userRole);
+                    continue outerloop;
                   }
-                }
 
-                existingUser.platformRoles[userPlatformRoleToUpdate].programs = existingUser.platformRoles[
-                  userPlatformRoleToUpdate
-                ].programs.map((eachProgram) => ObjectId(eachProgram));
+                  const currentRoleInfoIndex = existingUserProgramRoleMapping.findIndex(
+                    (pr) => pr.programId.toString() === programId.toString()
+                  );
 
-                updateQuery['platformRoles'] = existingUser.platformRoles;
+                  const newRoles = platform_role_array;
+
+                  if (userRole.programOperation === messageConstants.common.OVERRIDE_OPERATION) {
+                    if (currentRoleInfoIndex !== -1) {
+                      const currentRoles = existingUserProgramRoleMapping[currentRoleInfoIndex].roles;
+
+                      // Find roles to remove (exist in current but not in new)
+                      const rolesToRemove = currentRoles.filter((role) => !newRoles.includes(role));
+                      for (const role of rolesToRemove) {
+                        kafkaEventPayloads.push({
+                          userId: userProfile.id,
+                          username: userProfile.username,
+                          programId: programId,
+                          role,
+                          eventType: messageConstants.common.DELETE_EVENT_TYPE,
+                        });
+                      }
+
+                      // Find roles to add (exist in new but not in current)
+                      const rolesToAdd = newRoles.filter((role) => !currentRoles.includes(role));
+                      for (const role of rolesToAdd) {
```

Comment:
@borkarsaish65 please create functions for similar code.

Reply:
done